sdk/cosmos/azure_data_cosmos/src/retry_policies/client_retry_policy.rs

```diff
@@ -0,0 +1,1130 @@
+
+    /// Context information for routing the next retry attempt to a specific location
+    retry_context: Option<RetryContext>,
+
+    /// Clone of the original request used for retry decision logic
+    cosmos_request: Option<CosmosRequest>,
+
+    /// Underlying policy for handling resource throttling (429) with exponential backoff
+    throttling_retry: ResourceThrottleRetryPolicy,
+}
+
+impl ClientRetryPolicy {
+    /// Creates a new ClientRetryPolicy instance.
+    ///
+    /// # Summary
+    /// Initializes a retry policy that handles various failure scenarios including session token
+    /// mismatches, endpoint failures, service unavailability, and resource throttling. The policy
+    /// manages automatic endpoint discovery, multi-region failover, and coordinates with the
+    /// GlobalEndpointManager for routing decisions. It wraps a ResourceThrottleRetryPolicy for
+    /// handling 429 (TooManyRequests) responses with exponential backoff.
+    ///
+    /// # Arguments
+    /// * `global_endpoint_manager` - The endpoint manager for handling multi-region routing and failover
+    ///
+    /// # Returns
+    /// A new `ClientRetryPolicy` instance configured with default retry limits and throttling behavior
+    pub fn new(global_endpoint_manager: GlobalEndpointManager) -> Self {
+        Self {
+            global_endpoint_manager: Arc::new(global_endpoint_manager),
+            enable_endpoint_discovery: true,
+            failover_retry_count: 0,
+            session_token_retry_count: 0,
+            service_unavailable_retry_count: 0,
+            is_read_request: false,
+            can_use_multiple_write_locations: false,
+            is_multi_master_write_request: false,
+            location_endpoint: None,
+            retry_context: None,
+            cosmos_request: None,
+            throttling_retry: ResourceThrottleRetryPolicy::new(5, 200, 10),
+        }
+    }
+
+    /// Prepares a request before it is sent, configuring routing and endpoint selection.
+    ///
+    /// # Summary
+    /// Performs pre-flight setup for each request attempt by refreshing location cache,
+    /// determining request characteristics (read vs write, multi-master support), and
+    /// resolving the target endpoint based on retry context. Handles location-based routing
+    /// directives, including retry attempts that target specific location indices or the hub
+    /// endpoint. Clears previous routing context and configures the request with the
+    /// appropriate endpoint URL for the current attempt.
+    ///
+    /// # Arguments
+    /// * `request` - The mutable request to configure before sending
+    pub(crate) async fn before_send_request(&mut self, request: &mut CosmosRequest) {
+        self.cosmos_request = Some(request.clone());
+        let _stat = self
+            .global_endpoint_manager
+            .refresh_location_async(false)
+            .await;
+        self.is_read_request = request.is_read_only_request();
```

Comment:
If we do end up holding the whole request, we don't really need this value in a separate field. We can always read it out of the request.

Reply:
This is now refactored.